core/services/synchronization/telemetry_ingress_client.go

```diff
@@ -0,0 +1,156 @@
+package synchronization
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/service"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	pb "github.com/smartcontractkit/chainlink/core/services/synchronization/telem"
+	"github.com/smartcontractkit/chainlink/core/utils"
+
+	"github.com/smartcontractkit/wsrpc"
+	"github.com/smartcontractkit/wsrpc/examples/simple/keys"
+)
+
+// TelemetryIngressClient encapsulates all the functionality needed to
+// send telemetry to the ingress server using wsrpc
+type TelemetryIngressClient interface {
+	service.Service
+	Start() error
+	Close() error
+	Send(context.Context, []byte, common.Address)
+}
+
+type NoopTelemetryIngressClient struct{}
+
+func (NoopTelemetryIngressClient) Start() error                                 { return nil }
+func (NoopTelemetryIngressClient) Close() error                                 { return nil }
+func (NoopTelemetryIngressClient) Send(context.Context, []byte, common.Address) {}
+func (NoopTelemetryIngressClient) Healthy() error                               { return nil }
+func (NoopTelemetryIngressClient) Ready() error                                 { return nil }
+
+type telemetryIngressClient struct {
+	utils.StartStopOnce
+	url             *url.URL
+	ks              keystore.CSAKeystoreInterface
+	serverPubKeyHex string
+	wsrpcClient     pb.TelemClient
+	logging         bool
+
+	mu      *sync.RWMutex
+	isReady bool
+	wgDone  sync.WaitGroup
+	chDone  chan struct{}
+}
+
+// NewTelemetryIngressClient returns a client backed by wsrpc that
+// can send telemetry to the telemetry ingress server
+func NewTelemetryIngressClient(url *url.URL, serverPubKeyHex string, ks keystore.CSAKeystoreInterface, logging bool) TelemetryIngressClient {
+	return &telemetryIngressClient{
+		url:             url,
+		ks:              ks,
+		serverPubKeyHex: serverPubKeyHex,
+		logging:         logging,
+		mu:              new(sync.RWMutex),
+	}
+}
+
+// Start connects the wsrpc client to the telemetry ingress server
+func (tc *telemetryIngressClient) Start() error {
+	return tc.StartOnce("TelemetryIngressClient", func() error {
+		privkey, err := tc.getCSAPrivateKey()
+		if err != nil {
+			return err
+		}
+
+		tc.connect(privkey)
+
+		return nil
+	})
+}
+
+// Close disconnects the wsrpc client from the ingress server
+func (tc *telemetryIngressClient) Close() error {
+	return tc.StopOnce("TelemetryIngressClient", func() error {
+		close(tc.chDone)
+		tc.wgDone.Wait()
+		return nil
+	})
+}
+
+func (tc *telemetryIngressClient) connect(clientPrivKey []byte) {
+	tc.wgDone.Add(1)
+
+	go func() {
+		defer tc.wgDone.Done()
+
+		serverPubKey := keys.FromHex(tc.serverPubKeyHex)
+
+		conn, err := wsrpc.Dial(tc.url.String(), wsrpc.WithTransportCreds(clientPrivKey, serverPubKey))
```

Comment:
Does this handle reconnections?

Reply:
I believe it does, but will confirm with @jkongie

Reply:
Was able to verify the lib reconnects while testing manually